internal/filesystem/memdiff.go

```diff
@@ -0,0 +1,260 @@
+}
+
+func (mdfs *MemDiffFileSystem) Delete(ctx context.Context, path string) error {
+	// Remove from working tree if it exists there
+	delete(mdfs.workingTree, path)
+
+	// Add to deleted files
+	mdfs.deletedFiles[path] = struct{}{}
+	return nil
+}
+
+func (mdfs *MemDiffFileSystem) FileExists(ctx context.Context, path string) (bool, error) {
+	// Check if file is deleted
+	if _, exists := mdfs.deletedFiles[path]; exists {
+		return false, nil
+	}
+
+	// Check if file is in working tree
+	if _, exists := mdfs.workingTree[path]; exists {
+		return true, nil
+	}
+
+	// Fall back to base file system
+	return mdfs.baseFileSystem.FileExists(ctx, path)
+}
+
+func (mdfs *MemDiffFileSystem) IsDir(ctx context.Context, dir string) (bool, error) {
+	return mdfs.baseFileSystem.IsDir(ctx, dir)
+}
+
+func (mdfs *MemDiffFileSystem) ListDir(ctx context.Context, dir string) ([]string, error) {
+	baseFiles, err := mdfs.baseFileSystem.ListDir(ctx, dir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a map to track all files
+	allFiles := make(map[string]struct{})
+
+	// Add base files
+	for _, file := range baseFiles {
+		allFiles[file] = struct{}{}
+	}
+
+	// Add files from working tree that are in this directory
+	for path := range mdfs.workingTree {
+		if strings.HasPrefix(path, dir+"/") || path == dir {
+			allFiles[path] = struct{}{}
+		}
+	}
+
+	// Remove deleted files
+	for path := range mdfs.deletedFiles {
+		delete(allFiles, path)
+	}
+
+	// Convert back to slice
+	result := make([]string, 0, len(allFiles))
+	for file := range allFiles {
+		result = append(result, file)
+	}
+
+	return result, nil
+}
+
+// GetChangelist returns the changes tracked in this filesystem as a Changelist
+func (mdfs *MemDiffFileSystem) GetChangelist() git.Changelist {
+	return &memDiffChangelist{
+		workingTree:  mdfs.workingTree,
+		deletedFiles: mdfs.deletedFiles,
+	}
+}
+
+// memDiffChangelist implements the Changelist interface for MemDiffFileSystem
+type memDiffChangelist struct {
+	workingTree  map[string]string
+	deletedFiles map[string]struct{}
+}
+
+func (mdcl *memDiffChangelist) ForEachModified(fn func(path string, content string) error) error {
+	for path, content := range mdcl.workingTree {
+		if err := fn(path, content); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (mdcl *memDiffChangelist) ForEachDeleted(fn func(path string) error) error {
+	for path := range mdcl.deletedFiles {
+		if err := fn(path); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (mdcl *memDiffChangelist) IsModified(path string) bool {
+	_, exists := mdcl.workingTree[path]
+	return exists
+}
+
+func (mdcl *memDiffChangelist) IsDeleted(path string) bool {
+	_, exists := mdcl.deletedFiles[path]
+	return exists
+}
+
+func (mdcl *memDiffChangelist) IsEmpty() bool {
+	return len(mdcl.workingTree) == 0 && len(mdcl.deletedFiles) == 0
+}
+
+// githubReadOnlyFileSystem implements ReadOnlyFileSystem for GitHub repositories
+type githubReadOnlyFileSystem struct {
```

Comment:
`memdiff.go` isn't the right place for our github file system implementations. Either generalize the file name or move these implementations to another file like `github.go`